### DISCUSSION TITLE
如何使用docker部署一个机器人

### DISCUSSION THREAD

**Discussion by zhihui-Yu** [simple-robot/Discussions#60] [Q&A]
近期，个人想做一个机器人部署到k8s或者docker中。遇到一些问题来请教一下：
1. 如何在docker中运行？因为滑动验证器及短信校验让我对在linux中运行有了些担忧，并且没有找到相关教程
2. 如果运行了，是否会在长时间后掉线或者被强制退出

**Comment by ForteScarlet** [comment]
> 1. 如何在docker中运行？因为滑动验证器及短信校验让我对在linux中运行有了些担忧，并且没有找到相关教程

框架本身就是直接通过JVM运行的，任何允许的Java相关docker环境应该都能用，但是我本人并不熟悉docker因此无法解答。
其他讨论中偶尔也会有通过docker运行而产生疑问的，因此理论上docker部署是可行的。

但是docker环境对于验证确实有软肋，例如可能会因为docker环境下没有标准输入而无法进行验证等问题，这些问题也许能通过提前准备好 `device.json` 文件来解决，但也许也解决不了。我觉得如非必要，可以考虑直接使用Jar的方式部署。

> 2. 如果运行了，是否会在长时间后掉线或者被强制退出

对于qq机器人，[mirai组件](https://github.com/simple-robot/simbot-component-mirai)依赖于[mirai](https://github.com/mamoe/mirai)框架。mirai目前已经十分完善且社区活跃（大概），通常情况下不会出现长时间掉线或被强制退出的这类严重缺陷。当然，如果是账号重复登录导致的强制下线、账号被封禁等情况则另当别论。
> 也不排除遥远的未来腾讯突然大改QQ协议，或者mirai框架突然停止更新等不可预知的情况。但目前来看可能性十分微小。

**Reply by zhihui-Yu** [reply]
~对于使用jar部署，我尝试了，但是启动时由于linux环境无法做验证码+短信验证，所以直接报错了~
![image](https://user-images.githubusercontent.com/54210717/215650448-69d5269b-c76f-48ac-8a3b-273cc44614cd.png)

~不知道是不是需要特殊的配置或者操作在linux环境中。~

对于我所提的问题，尝试了在配置文件中设置`simbot.auto-start-bots=false` 以及添加jvm参数 -Dmirai.no-desktop后运行，下载了[apk](https://github.com/KasukuSakura/mirai-login-solver-sakura/releases/tag/v0.0.8)，输入控制台显示的连接后获取到了ticket，输入在console上后，显示如下：
![image](https://user-images.githubusercontent.com/54210717/215679079-90de9c4f-2f38-4584-812f-a5aef315dbcb.png)

修改protocol为`ANDROID_PAD` 依旧不行 (配置文件中添加了`simbot.component.mirai.protocol=ANDROID_PAD`), 重启后显示如下
![image](https://user-images.githubusercontent.com/54210717/215684447-dd04eeed-2662-4b1f-b49e-781a22f28a21.png)


在[搜索](https://mirai.mamoe.net/topic/1668/%E6%8F%90%E7%A4%BAqq%E7%89%88%E6%9C%AC%E8%BF%87%E4%BD%8E)后,说了被风控了，可能要换号了。 

基于上述，如果qq和微信风控都如此严重，那机器人如何能长久运行便是一个问题了， 不知是否有办法解决呢

此外，验证器相关的地方我有些模糊，不知道如何让应用变成扫描登入呢(在控制台显示二维码)，我暂时没有找到相关文档

**Reply by ForteScarlet** [reply]
QQ版本过低的问题在mirai的 [v2.13.3](https://github.com/mamoe/mirai/releases/tag/v2.13.3) 版本解决过一次（对应组件版本 [v3.0.0.0-M1](https://github.com/simple-robot/simbot-component-mirai/releases/tag/v3.0.0.0-M1)），如果不是此版本可以考虑更新版本后再尝试。

> qq和微信风控都如此严重，那机器人如何能长久运行便是一个问题了， 不知是否有办法解决呢

~~这种问题你只能去找老马解决了（~~  tx的风控就算是普通用户正常使用APP都会存在，更何况是可能频繁变更设备、登录地区的bot程序。不过目前来讲我个人的bot已经相对稳定的运行了至少两年，最长持续运行也能有数月，只要能确保不会被举报、不会频繁更换设备信息、登陆地点（以及考虑再冲个QQ会员）一般不会有大问题。

_归根结底，mirai或其他相关框架都并非TX官方认可的东西，这本来就是非常规应用方式，官方有充足的理由进行非常严格的风控。_

扫描登录的问题，参考 https://github.com/mamoe/mirai/issues/1281 ，mirai目前似乎尚未支持。

**Reply by zhihui-Yu** [reply]
我当前的版本是 love.forte.simple-robot:component-mirai:2.5.1, 看mvn repo，应该是最新的一个了。

关于扫描登入，我看那个ticket依旧是open，看起来很久没有变化了。（没戏）

请问一下，你所运行的平台(win/linux/mac)哪个呢，因为我云服务装linux (还有其他用处所以不换其他os)，所以一直尝试linux环境。
我在想如果在win上面登入后，将cache，device等信息复制到linux环境的jar根目录上，是否可以跳过验证呢。

**Reply by ForteScarlet** [reply]
> 我当前的版本是 love.forte.simple-robot:component-mirai:2.5.1

v2版本的simbot在目前v3发布的情况下，将会很快停止更新，`v2.5.1` 对应的mirai版本是 `v2.13.2` 而不是 `v2.13.3`，如果需要可以考虑手动排除并自行引入 `2.13.3` 版本的 `mirai-core`。

> 请问一下，你所运行的平台(win/linux/mac)哪个呢

我是用的是阿里云的Linux，只不过是上古的centOS系统，但是没什么影响。

> 将cache，device等信息复制到linux环境的jar根目录上，是否可以跳过验证呢

理论上可行，但是并不是100%，因为放到服务器运行相当于异地登陆。但是在Linux上应当依旧可以进行验证，毕竟验证都是控制台完成，并没有什么图形化界面，不应影响什么。
